### PR TITLE
Upgrade FasterTransformer base on nccl_dependent_refine branch

### PR DESCRIPTION
**Pull request by FrostML**
### PR types

Others

### PR changes

Models & APIs

### Description

Upgrade FasterTransformer base on nccl_dependent_refine branch.
Done: 
* [Decoding] Transformer @FrostML 
  * pre-normalization
  * beam search v1
  * beam search v2
  * sampling (topk/topp)
* [Decoding] GPT @FrostML 
  * single cards
* [Decoding] BART @FrostML 
  * post-normalization
  * beam search v1
  * beam search v2
  * sampling (topk/topp)
* [Decoding] UnifiedTransformer @FrostML 
  * pre-normalization
  * decoding strategies are the same with above model.
* [Decoding] UNIMOText @FrostML 
  * post-normalization
  * decoding strategies are the same with above model.
* [Decoder] Pre-normalization @FrostML 
* [Decoder] Post-normalization @gongel 
* [Decoder] Integrate different attention. @gongel 
* [Decoder] Decoder OP. @gongel 
* [Encoder] Post-/Pre-Norm @tianxin1860

### CODE REVIEW

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/open_decoder.h [hunk 1]
```diff
@@ -203,6 +203,8 @@ class OpenDecoder {
     l_parallel_param_ = param;
   }
 
+  void initialize_stream(cudaStream_t stream) { param_.stream = stream; }
```
Comment (FrostML) [inline paddlenlp/ops/patches/FasterTransformer/fastertransformer/open_decoder.h]: @gongel 这个函数删除吧，在新版上没有使用了

File: paddlenlp/ops/faster_transformer/src/fusion_bart_decoding_op.cu [hunk 2]
```diff
@@ -257,10 +258,11 @@ std::vector<paddle::Tensor> bart_decoding_kernel(
         start_id_,
         end_id_,
         beam_search_diversity_rate_,
-        false,
-        false,
+        false, /*is_fuse_topk_softMax*/
```
Comment (guoshengCS) [inline paddlenlp/ops/faster_transformer/src/fusion_bart_decoding_op.cu]: `is_fuse_topk_softMax`这个默认打开吧，和FT保持一致，也为了性能考虑。要验证下正确性

File: examples/machine_translation/transformer/faster_transformer/encoder_decoding_predict.py [hunk 3]
```diff
@@ -41,10 +41,12 @@ def parse_args():
...
         default="beam_search",
         type=str,
-        choices=["beam_search", "topk_sampling", "topp_sampling"],
+        choices=[
+            "beam_search", "beam_search_v2", "topk_sampling", "topp_sampling"
+        ],
         help="Decoding strategy. Can be one of ['beam_search', 'topk_sampling', 'topp_sampling']. "
     )
-    parser.add_argument("--beam_size", default=4, type=int, help="Beam size. ")
+    parser.add_argument("--beam_size", default=5, type=int, help="Beam size. ")
```
Comment (guoshengCS) [inline examples/machine_translation/transformer/faster_transformer/encoder_decoding_predict.py]: 如果4能用fuse_topk_softmax而5不能，速度相差比较明显，默认还是用4吧

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/open_decoder.h (hunk 1 above)
Comment (gongel) [reply paddlenlp/ops/patches/FasterTransformer/fastertransformer/open_decoder.h]: DONE

File: paddlenlp/ops/faster_transformer/sample/decoder_sample.py [hunk 4]
```diff
@@ -104,12 +136,19 @@ def do_predict(args):
...
-                self_cache=self_cache,
-                mem_cache=mem_cache)
+                self_cache_key=self_cache_key,
+                self_cache_value=self_cache_value,
+                mem_cache=mem_cache,
+                step=0,
+                memory_hidden_dim=args.d_model,
+                is_fuse_qkv=False)
+        paddle.device.cuda.synchronize()
+        print(dec_output)
```
Comment (guoshengCS) [inline paddlenlp/ops/faster_transformer/sample/decoder_sample.py]: 注意去掉调试代码

File: paddlenlp/ops/faster_transformer/sample/encoder_decoder_sample.py [hunk 5]
```diff
@@ -79,11 +94,14 @@ def do_predict(args):
         dtype='int64')
 
     with paddle.no_grad():
-        for i in range(100):
+        for i in range(1):
             # For warmup. 
-            if 50 == i:
+            if 0 == i:
```
Comment (guoshengCS) [inline paddlenlp/ops/faster_transformer/sample/encoder_decoder_sample.py]: 是否在调试后忘记改回去

File: paddlenlp/ops/faster_transformer/sample/gpt_sample.py [hunk 6]
```diff
@@ -52,7 +52,7 @@ def parse_args():
         "--batch_size", default=4, type=int, help="Batch size. ")
     parser.add_argument(
         "--topk",
-        default=4,
+        default=1,
```
Comment (guoshengCS) [inline paddlenlp/ops/faster_transformer/sample/gpt_sample.py]: 是否在调试后忘记改回去

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/cuda_kernels.cu [hunk 7]
```diff
@@ -27,16 +27,18 @@ __global__ void update_logits_kernel(T* logits,
...
+  const T MAX_T_VAL = (sizeof(T) == 2) ? HALF_FLT_MAX : FLT_MAX;
+
+  float max_val = -FLT_MAX;
   __shared__ float s_max_val;
   __shared__ float s_sum_val;
 
   for (int tid = threadIdx.x; tid < n; tid += blockDim.x) {
     if (finish)
-      logits[offset + tid] = (tid == end_id) ? FLT_MAX : -1 * FLT_MAX;
+      logits[offset + tid] = (tid == end_id) ? MAX_T_VAL : -MAX_T_VAL;
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/cuda_kernels.cu:38]: 这里是为什么要修改呢，看beam search v1的logits都是使用的float

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/cuda_kernels.cu [hunk 8]
```diff
@@ -69,10 +73,29 @@ void update_logits_v2(float* logits,
...
 
+template void update_logits_v2(float* logits,
+                               const float* bias,
+                               const int end_id,
+                               const bool* finished,
+                               const int m,
+                               const int n,
+                               cudaStream_t stream);
+
+template void update_logits_v2(half* logits,
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/cuda_kernels.cu:88]: 注意beam search v1的logits是否都是float

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/decoding_kernels.cu [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+
+template <typename T, bool ALIVE = false>
+__global__ void init_kernel_v2(bool* finished,
+                               bool* alive_finished,
+                               int* sequence_length,
+                               int* word_ids,
+                               T* cum_log_probs,
+                               const int sentence_id,
+                               const int beam_width,
+                               const int batch_size) {
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/decoding_kernels.cu:28]: 最好也和cuda_kernels.cu一样使用move，这样能够看出来diff

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/decoding_beamsearch.h [hunk 10]
```diff
@@ -0,0 +1,1115 @@
...
+#include "fastertransformer/open_decoder.h"
+#include "fastertransformer/utils/allocator.h"
+#include "fastertransformer/utils/arguments.h"
+#include "fastertransformer/utils/common.h"
+#include "fastertransformer/utils/functions.h"
+
+namespace fastertransformer {
+
+template <OperationType OpType_>
+class DecodingBeamsearch {
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/fastertransformer/decoding_beamsearch.h:34]: 最好也和cuda_kernels.cu一样使用move，这样能够看出来diff

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu [hunk 11]
```diff
@@ -397,16 +437,19 @@ __global__ void topk_stage_2_opt3(const int* __restrict topk_tmp_id_buf,
...
     int* topk_tmp_id_buf,
     T* topk_tmp_val_buf,
+    const bool* finished,
     const int k,
     const int vocab_size,
-    const T diversity_rate) {  // diversity_rate only works when
-                               // BLOCKS_PER_BEAM_ is 1
+    const T
+        diversity_rate,  // diversity_rate only works when BLOCKS_PER_BEAM_ is 1
+    const int end_id) {
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu:452]: finished和end_id如果没有使用到的话就去掉吧

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu [hunk 12]
```diff
@@ -776,15 +907,15 @@ __global__ void topk_stage_2_opt3_update(const int* __restrict topk_tmp_id_buf,
         if (step == 1) {  // init output
           for (int i = 0; i < beam_width; i++) {
             output_word_ids[i] = end_id;
-            output_cum_log_probs[i] = -MAX_T_VAL;
+            output_cum_log_probs[i] = -1e20f;
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu:910]: 最好再测试确认下online_softmax中batch_topk_update_kernel仍使用`output_cum_log_probs[i] = -MAX_T_VAL`在fp16时是否会有问题，在output_cum_log_probs改为float后里面还会和TopKFinish中的T的值相互转换比较

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu [hunk 13]
```diff
@@ -1141,11 +1306,17 @@ void topK_sampling_kernel_kernelLauncher(void* workspace,
                                          bool* finished_buf,
                                          int random_num,
                                          DecodingSamplingArguments args,
-                                         cudaStream_t stream) {
+                                         cudaStream_t stream,
+                                         const int batch_size) {
+  // TODO: check stochasticity.
   std::minstd_rand engine;
   int seed = std::random_device()();
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu]: 这个patch现在还有调用吗

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu [hunk 14]
```diff
@@ -1497,21 +2011,24 @@ void topK_topP_sampling_kernel_kernelLauncher(void* workspace,
                                               const T* logits,
                                               const int random_num,
                                               DecodingSamplingArguments& args,
-                                              cudaStream_t stream) {
+                                              cudaStream_t stream,
+                                              const int batch_size) {
+  // TODO: verify stochasticity
   std::minstd_rand engine;
   int seed = std::random_device()();
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu]: 同上，这里的patch是否还有调用，如果没有调用的话这个文件的patch能否改成append的，现在确实不太好区分哪些是patch的内容

File: paddlenlp/ops/faster_transformer/sample/decoder_sample.py (hunk 4 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/sample/decoder_sample.py]: @gongel

File: paddlenlp/ops/faster_transformer/sample/encoder_decoder_sample.py (hunk 5 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/sample/encoder_decoder_sample.py]: @gongel

File: paddlenlp/ops/faster_transformer/sample/decoder_sample.py (hunk 4 above)
Comment (gongel) [reply paddlenlp/ops/faster_transformer/sample/decoder_sample.py]: done and thx

File: paddlenlp/ops/faster_transformer/sample/encoder_decoder_sample.py (hunk 5 above)
Comment (gongel) [reply paddlenlp/ops/faster_transformer/sample/encoder_decoder_sample.py]: done and thx

File: examples/machine_translation/transformer/faster_transformer/encoder_decoding_predict.py (hunk 3 above)
Comment (FrostML) [reply examples/machine_translation/transformer/faster_transformer/encoder_decoding_predict.py]: Done. Thanks.

File: paddlenlp/ops/faster_transformer/sample/gpt_sample.py (hunk 6 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/sample/gpt_sample.py]: Done. Thanks.

File: paddlenlp/ops/faster_transformer/src/fusion_bart_decoding_op.cu (hunk 2 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/src/fusion_bart_decoding_op.cu]: Done. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/cuda_kernels.cu (hunk 7 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/cuda_kernels.cu:38]: v1 这里使用的是 `float* logits`，v2 直接是用的 template 指定的类型，可以为 float16.

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/cuda_kernels.cu (hunk 8 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/cuda_kernels.cu:88]: v2 可以为 template 指定类型，即，可为 float16.

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/decoding_kernels.cu (hunk 9 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/decoding_kernels.cu:28]: 确实是用的 move，不过可能改动比较多，github 自动视作创建新的文件了。

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/decoding_beamsearch.h (hunk 10 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/fastertransformer/decoding_beamsearch.h:34]: 确实是用的 move，不过可能改动比较多，github 自动视作创建新的文件了。

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu (hunk 12 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu:910]: `output_cum_log_probs` 在 4.0 上类型为 float，不再是 float16，所以这里这么修改，经测试，对结果没有影响。

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu (hunk 13 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu]: 已移除

File: paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu (hunk 14 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/topk_kernels.cu]: 已移除

**Review by guoshengCS — APPROVED** [review]
